File: Tool/PreUtil/src/output/javascript/hard/ContainsDuplicateIii.js
/**https://leetcode.com/problems/contains-duplicate-iii/ */
//You are given an integer array nums and two integers indexDiff and valueDiff.
//Find a pair of indices (i, j) such that:
//i != j,
//abs(i - j) <= indexDiff.
//abs(nums[i] - nums[j]) <= valueDiff, and
//Return true if such pair exists or false otherwise.


/**
 * @param {number[]} nums
 * @param {number} indexDiff
 * @param {number} valueDiff
 * @return {boolean}
 */
var containsNearbyAlmostDuplicate = function(nums, indexDiff, valueDiff) {
    let n = nums.length;
    let dp = [];
    for(let i = 0; i <= n; i++){
        dp.push([]);
        for(let j = 0; j <= indexDiff; j++){
            dp[i].push([]);
            for(let k = 0; k <= valueDiff; k++){
                dp[i][j].push(false);
            }
        }
    }
    for(let i = 1; i <= n; i++){
        for(let j = 1; j <= indexDiff; j++){
            for(let k = 1; k <= valueDiff; k++){
                if(i - j >= 0 && Math.abs(nums[i - 1] - nums[i - j - 1]) <= k){
                    dp[i][j][k] = true;
                }
                else{
                    dp[i][j][k] = dp[i - 1][j][k];
                }
            }
        }
    }
    return dp[n][indexDiff][valueDiff];
    
};

    

File: Tool/PreUtil/src/output/java/hard/ContainsDuplicateIii.java
/**https://leetcode.com/problems/contains-duplicate-iii/ */
//You are given an integer array nums and two integers indexDiff and valueDiff.
//Find a pair of indices (i, j) such that:
//i != j,
//abs(i - j) <= indexDiff.
//abs(nums[i] - nums[j]) <= valueDiff, and
//Return true if such pair exists or false otherwise.


class ContainsDuplicateIii {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int indexDiff, int valueDiff) {
        int n = nums.length;
        TreeSet<Long> set = new TreeSet<>();
        for (int i = 0; i < n; i++) {
            Long ceiling = set.ceiling((long) nums[i] - valueDiff);
            if (ceiling != null && ceiling <= (long) nums[i] + valueDiff) {
                return true;
            }
            set.add((long) nums[i]);
            if (i >= indexDiff) {
                set.remove((long) nums[i - indexDiff]);
            }
        }
        return false;
        
    }
}
     
    